File: parser/BrightScriptEventListener.js
const BrightScriptEventGenerator = require('./BrightScriptEventGenerator'),
    BrightScriptParser = require('./antlr/BrightScriptParser').BrightScriptParser;

const FunctionType = {
    Void: 'Void',
    Dynamic: 'Dynamic'
};

const isFunction = (context) => isSub(context) || context.ruleIndex === BrightScriptParser.RULE_anonymousFunctionDeclaration || context.ruleIndex === BrightScriptParser.RULE_functionDeclaration;
const isSub = (context) => context.ruleIndex === BrightScriptParser.RULE_anonymousSubDeclaration || context.ruleIndex === BrightScriptParser.RULE_subDeclaration;
const typeText = (token, context) => {
    if (isSub(context) || (token && token.getText().toUpperCase() === 'VOID')) {
       return FunctionType.Void;
    }
    
    if (token != null) {
        let text = token.getText(),
            first = text[0].toUpperCase();
            
        return `${first}${text.slice(1).toLowerCase()}`;
    }

    return FunctionType.Dynamic;
};

class BrightScriptEventListener extends BrightScriptEventGenerator {
    constructor(emitter) {
        super(emitter);

        this.functionNames = [];
    }

    enterFunctionDeclaration(context) {
        this.checkDeclaration(context);
    }

    enterSubDeclaration(context) {
        this.checkDeclaration(context);
    }

    exitReturnStatement(context) {
        let functionContext = context;

        while (functionContext.parentCtx != null && !isFunction(functionContext)) {
            functionContext = functionContext.parentCtx;
        }

        // We check isFunction again because we may have traversed all the way to
        // the top of the file. If that happens we need to bail.
        if (functionContext == null || !isFunction(functionContext)) {
            return;
        }

        const expression = context.children[1];
        const baseTypeToken = functionContext.children.find(child => child.ruleIndex === BrightScriptParser.RULE_baseType);
        const baseType = typeText(baseTypeToken, functionContext);

        if (baseType === FunctionType.Void && expression != null) {
            context.parser.notifyErrorListeners(`return value unexpected for function with return type of '${baseType}'`, context.start);
        }
        else if (baseType !== FunctionType.Void && expression == null) {
            context.parser.notifyErrorListeners(`return value expected for function with return type of '${baseType}'`, context.start);
        }
    }

    checkDeclaration(context) {
        const functionExists = name => this.functionNames.some(funcName => name.toUpperCase() === funcName.toUpperCase());

        context.children.forEach(child => {
            if (child.ruleIndex === BrightScriptParser.RULE_untypedIdentifier) {
                let name = child.getText();

                if (functionExists(name)) {
                    child.parser.notifyErrorListeners(`function '${name}' is declared multiple times`, context.start);
                }
                else {
                    this.functionNames.push(name);
                }
            }

            return;
        });
    }
}

module.exports = BrightScriptEventListener;

File: src/cpp/BrightscriptEventListener.cpp
#include "BrightScriptEventListener.h"
#include "Node.h"

using namespace std;
using namespace emscripten;
using namespace antlr4;

static inline char string_toupper_functional(char c)
{
    return toupper(c);
}

static inline string string_upper(const string &str)
{
    //TODO - replace strcopy
    string strcopy(str.size(), 0);
    transform(str.begin(), str.end(), strcopy.begin(), string_toupper_functional);
    return strcopy;
}

bool isSub(ParserRuleContext *ctx)
{
    return ctx->getRuleIndex() == BrightScriptParser::RuleAnonymousSubDeclaration ||
           ctx->getRuleIndex() == BrightScriptParser::RuleSubDeclaration;
}

TreeNode buildTreeFromContext(ParserRuleContext *ctx, BrightScriptParser *parser)
{
    vector<TreeNode> children = {};
    for (auto child : ctx->children)
    {
        if (ParserRuleContext *childCtx = dynamic_cast<ParserRuleContext *>(child))
        {
            children.push_back(buildTreeFromContext(childCtx, parser));
        }
    }
    return TreeNode{
        Node{parser->getRuleNames()[ctx->getRuleIndex()], ctx->getText()},
        children};
}

bool isFunction(ParserRuleContext *ctx)
{
    return isSub(ctx) ||
           ctx->getRuleIndex() == BrightScriptParser::RuleAnonymousFunctionDeclaration ||
           ctx->getRuleIndex() == BrightScriptParser::RuleFunctionDeclaration;
}

ParserRuleContext *getParentFunctionContext(ParserRuleContext *ctx)
{
    if (ParserRuleContext *parent = dynamic_cast<ParserRuleContext *>(ctx->parent))
    {
        return isFunction(parent) ? parent : getParentFunctionContext(parent);
    }
    return nullptr;
}

void BrightScriptEventListener::enterFunctionDeclaration(BrightScriptParser::FunctionDeclarationContext *ctx)
{
    checkDeclaration(ctx->untypedIdentifier());
}

void BrightScriptEventListener::enterSubDeclaration(BrightScriptParser::SubDeclarationContext *ctx)
{
    checkDeclaration(ctx->untypedIdentifier());
}

void BrightScriptEventListener::exitReturnStatement(BrightScriptParser::ReturnStatementContext *ctx)
{
    auto functionParent = getParentFunctionContext(ctx);
    if (functionParent == nullptr)
    {
        return;
    }

    // if we can't figure out the type, it's
    // probably dynamic
    // TODO - Replace string constants with enums for performance
    string type = "DYNAMIC";
    if (isSub(functionParent))
    {
        type = "VOID";
    }
    else
    {
        for (auto child : functionParent->children)
        {
            if (BrightScriptParser::BaseTypeContext *childCtx = dynamic_cast<BrightScriptParser::BaseTypeContext *>(child))
            {
                type = string_upper(childCtx->getText());
            }
        }
    }

    bool hasExpression = ctx->children.size() > 1;
    if (type == "VOID" && hasExpression)
    {
        // Parser instance isn't available on context in C++ target. Need to figure out a different way to do this.
        //ctx->parser->notifyErrorListeners(ctx->start, "Unexpected return value for function with return type " + type, nullptr);
    }
    else if (type != "VOID" && !hasExpression)
    {
        //ctx->parser->notifyErrorListeners(ctx->start, "Expected return value for function with return type " + type, nullptr);
    }
}

bool BrightScriptEventListener::functionNameExists(string nameToCheck)
{
    for (auto name : functionNames)
    {
        if (name == nameToCheck)
        {
            return true;
        }
    }
    return false;
}

void BrightScriptEventListener::checkDeclaration(BrightScriptParser::UntypedIdentifierContext *ctx)
{
    auto name = string_upper(ctx->getText());
    if (functionNameExists(name))
    {
        //ctx->parser->notifyErrorListeners(ctx->start, "function \"" + name + "\" is declared multiple times", nullptr);
    }
    else
    {
        functionNames.push_back(name);
    }
}
